### PR TITLE
Frogger game task

### PR DESCRIPTION
**Pull request by iva-stasia**
### Frogger game

[Demo](https://iva-stasia.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/iva-stasia/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
In addition to the bot's suggestions above, please post a link to the app demo in Students' chat, collect feedback and fix bugs if any based on that. When user tests are done post here a comment on that.

**Comment by iva-stasia** [comment]
Checked the code for common mistakes. It seems everything is fine.

However, in the first version of the demo, the player could move even when the modal windows weren't closed. I decided to fix it and after that I asked students in the chat to test my demo. No bugs were found.

File: submissions/iva-stasia/frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        this.x = INIT_VAL.enemyX;
+    };
+
+    this.checkCollision();
+};
+
+Enemy.prototype.checkCollision = function() {
+    if (player.x < this.x + INIT_VAL.collision &&
+        player.x + INIT_VAL.collision > this.x &&
+        player.y < this.y + INIT_VAL.collision &&
```
Comment (OleksiyRudenko) [inline submissions/iva-stasia/frogger-game/app.js:76]: Please find a requirement among those posted/linked by bot that addresses this case, post it as a comment and also fix the code accordingly.

Find other fragments where the same requirement is not met and fix them as well.

File: submissions/iva-stasia/frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+Enemy.prototype.render = function() {
+    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+const allEnemies = [];
+
+INIT_VAL.enemyY.forEach(y => {
+    let enemySpeed = Math.round(Math.random() * INIT_VAL.enemyMaxspeed + INIT_VAL.enemyMinSpeed);
+    let enemy = new Enemy(INIT_VAL.enemyX, y, enemySpeed);
```
Comment (OleksiyRudenko) [inline submissions/iva-stasia/frogger-game/app.js:90]: Why `let`?

File: submissions/iva-stasia/frogger-game/app.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        winMessage.classList.add('active_message');
+    };
+};
+
+const player = new Player(INIT_VAL.playerX, INIT_VAL.playerY);
+
+Player.prototype.handleInput = function(allowedKeys) {
+    if (!lostMessage.classList.contains('active_message')
+        && !winMessage.classList.contains('active_message')
+        && playerOptionDiv.classList.contains('hide')) {
```
Comment (OleksiyRudenko) [inline submissions/iva-stasia/frogger-game/app.js:143]: Making game objects dependant on environment is not a good practice.
The better practice would be to provide class with some external functions of interacting with the excution environment.
You do not need to fix this in this task.
Keep this principle in mind.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@iva-stasia good job!
One important class of typical problems to fix, one learning to take.

File: submissions/iva-stasia/frogger-game/app.js (hunk 1 above)
Comment (iva-stasia) [reply submissions/iva-stasia/frogger-game/app.js:76]: Sorry, I'm not sure I understood correctly.
Do you mean this?

> Classes do not refer to any global variables, like global variable player, which is an instance of Player class (referring to global constants and globals provided by the gaming platform like Resources is OK); Hint: pass instance of a game object (or objects) as an argument to other game objects they need to interact with.

If so, at the moment, I don't understand how I can check for enemy and player collision without specifying either one or the other. Even with a hint :(
Can I create a separate function that does not inherit anything?
I will try to understand what the hint means or ask the students for help.

Or does it mean that I can pass the player as an argument? In this way:
<pre><code>Enemy.prototype.checkCollision = function(player) {}
</code></pre>

File: submissions/iva-stasia/frogger-game/app.js (hunk 2 above)
Comment (iva-stasia) [reply submissions/iva-stasia/frogger-game/app.js:90]: When I wrote this code, I thought that inside the methods that execute a function for each element of the array, you need to declare variables using <code>let</code>. Now I already know that this is not so. I'll fix it.

Comment (iva-stasia) [reply submissions/iva-stasia/frogger-game/app.js]: :white_check_mark:

File: submissions/iva-stasia/frogger-game/app.js (hunk 1 above)
Comment (iva-stasia) [reply submissions/iva-stasia/frogger-game/app.js:76]: :white_check_mark:

If this solution doesn't work, I have another one. But it was suggested to me in the students' chat, so first I want to try it my way.

File: submissions/iva-stasia/frogger-game/app.js (hunk 3 above)
Comment (iva-stasia) [reply submissions/iva-stasia/frogger-game/app.js:142]: Thank you for this note.

File: submissions/iva-stasia/frogger-game/app.js (hunk 1 above)
Comment (OleksiyRudenko) [reply submissions/iva-stasia/frogger-game/app.js:76]: You still refer to a global variable within class methods definition. Just in different place.
The requirements offer a hint on how to resolve this issue.

Comment (iva-stasia) [reply submissions/iva-stasia/frogger-game/app.js]: I hope I finally figured it out. Also added calculation of initial positions of enemies.

**Review by OleksiyRudenko — APPROVED** [review]
@iva-stasia great job!

**Pull request merged**